experimental/go/internalgsdk.go

```diff
@@ -0,0 +1,262 @@
+	cachedScheduledMaintenanceUtc time.Time
+	startInternalRan              bool
+	// startInternalMutex is used to ensure that the startCore() method is only called once
+	startInternalMutex *sync.Mutex
+	debugLogs          bool
+}
+
+// newInternalGsdk creates a new instance of the internalGsdk struct
+func newInternalGsdk() *internalGsdk {
+	return &internalGsdk{
+		currentGameState:              "",
+		transitionToActiveEvent:       make(chan interface{}),
+		transitionToActiveEventClosed: false,
+		signalHeartbeatEvent:          make(chan interface{}),
+		configuration:                 nil,
+		configMap:                     nil,
+		healthCallback:                nil,
+		maintenanceCallback:           nil,
+		shutdownCallback:              nil,
+		connectedPlayers:              nil,
+		initialPlayers:                nil,
+		startInternalRan:              false,
+		startInternalMutex:            &sync.Mutex{},
+	}
+}
+
+// startInternal initializes the core GSDK functionality
+func (i *internalGsdk) startInternal() {
+	i.startInternalMutex.Lock()
+	defer i.startInternalMutex.Unlock()
+	if i.startInternalRan {
+		return
+	}
+
+	i.currentGameState = GameStateInitializing
+	var err error
+	if i.configuration == nil {
+		i.configuration, err = i.getConfiguration()
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	if i.configMap == nil {
+		i.configMap = i.createConfigMap(i.configuration)
+	}
+
+	heartbeatEndpoint := i.configMap[HeartbeatEndpointKey]
+	serverId := i.configMap[ServerIdKey]
+
+	initializeLogger(i.configMap[LogFolderKey])
+
+	logInfo(fmt.Sprintf("VM Agent Endpoint %s", heartbeatEndpoint))
+	logInfo(fmt.Sprintf("Instance Id %s", serverId))
+
+	// run heartbeats on a different goroutine
+	go i.heartbeat(fmt.Sprintf("http://%s/v1/sessionHosts/%s/heartbeats", heartbeatEndpoint, serverId))
+	i.startInternalRan = true
+}
+
+func (i *internalGsdk) heartbeat(heartbeatEndpoint string) {
+	// if we get a signal for a state transition, we'll heartbeat
+	// else, we just heartbeat every second
+	for {
+		select {
+		case <-i.signalHeartbeatEvent:
+			if i.debugLogs {
+				logInfo("State transition signaled an early heartbeat")
+			}
+			i.sendHeartbeat(heartbeatEndpoint)
+		case <-time.After(time.Second * 1):
+			i.sendHeartbeat(heartbeatEndpoint)
+		}
+	}
+}
+
+func (i *internalGsdk) sendHeartbeat(heartbeatEndpoint string) {
+	gameHealth := false
+	if i.healthCallback != nil {
+		gameHealth = i.healthCallback()
+	}
+	currentGameHealth := "Healthy"
+	if !gameHealth {
+		currentGameHealth = "Unhealthy"
+	}
+	hrr := HeartbeatRequest{
+		CurrentGameState:  i.currentGameState,
+		CurrentGameHealth: currentGameHealth,
+		CurrentPlayers:    i.connectedPlayers,
+	}
+
+	postBody, _ := json.Marshal(hrr)
+	requestBody := bytes.NewBuffer(postBody)
+
+	resp, err := http.Post(heartbeatEndpoint, "application/json", requestBody)
+	if err != nil {
+		logWarn(fmt.Sprintf("Error sending heartbeat %s", err))
+		return
+	}
+
+	defer resp.Body.Close()
```

Comment:
Should we be checking the status code on the response here before examining the body?

Reply:
great catch, thanks! adding it